tensorboard/plugins/pr_curve/pr_curves_plugin.py

```diff
@@ -0,0 +1,159 @@
+    """
+    runs = request.args.getlist('run')
+    tag = request.args.get('tag')
+
+    response_mapping = {}
+    for run in runs:
+      try:
+        tensor_events = self._multiplexer.Tensors(run, tag)
+      except KeyError:
+        return http_util.Respond(
+            request,
+            'No PR curves could be fetched for run %r and tag %r' % (run, tag),
+            'text/plain',
+            400)
+      
+      response_mapping[run] = [
+          self._process_tensor_event(e) for e in tensor_events]
+
+    # We convert the tensor data into a JSON-able response.
+    return http_util.Respond(request, response_mapping, 'application/json')
+
+  @wrappers.Request.application
+  def tags_route(self, request):
+    """A route (HTTP handler) that returns a response with tags.
+    Returns:
+      A response that contains a JSON object. The keys of the object
+      are all the runs. Each run is mapped to a (potentially empty)
+      list of all tags that are relevant to this plugin.
+    """
+    all_runs = self._multiplexer.PluginRunToTagToContent(
+        PrCurvesPlugin.plugin_name)
+    response = {
+        run: list(tag_to_content.keys())
+             for (run, tag_to_content) in all_runs.items()
+    }
+    return http_util.Respond(request, response, 'application/json')
+
+  @wrappers.Request.application
+  def steps_per_run_route(self, request):
+    """Gets a dict mapping run to a list of numeric steps.
+
+    Returns:
+      A dict with string keys (all runs with PR curve data). The values of the
+      dict are lists of step values (ints) that should be used for the slider
+      for that run.
+    """
+    all_runs = self._multiplexer.PluginRunToTagToContent(
+        PrCurvesPlugin.plugin_name)
+    response = {}
+
+    # Compute the max step exhibited by any tag for each run. If a run lacks
+    # data, exclude it from the mapping.
+    for run, tag_to_content in all_runs.items():
+      if not tag_to_content:
+        # This run lacks data for this plugin.
+        continue
+
+      # Just use the list of tensor events for any of the tags to determine the
+      # steps to list for the run. The steps are often the same across tags for
+      # each run, albeit the user may elect to sample certain tags differently
+      # within the same run. If the latter occurs, TensorBoard will show the
+      # actual step of each tag atop the card for the tag.
+      tensor_events = self._multiplexer.Tensors(
+          run, list(tag_to_content.keys())[0])
+      response[run] = [e.step for e in tensor_events]
+    return http_util.Respond(request, response, 'application/json')
+
+  def get_plugin_apps(self):
+    """Gets all routes offered by the plugin.
+
+    Returns:
+      A dictionary mapping URL path to route that handles it.
+    """
+    return {
+        '/tags': self.tags_route,
+        '/pr_curves': self.pr_curves_route,
+        '/steps_per_run': self.steps_per_run_route,
```

Comment:
As with the others, can we call this `/available_steps` or `/run_to_available_steps` or similar? (For consistency, `/available_steps` (or similar) is probably better; our `/tags` route gives us an object that we call `runToTag` on the frontend, etc.)